ext/rugged/rugged_tree.c

```diff
@@ -93,6 +93,63 @@ static VALUE rb_git_tree_entrycount(VALUE self)
 	return INT2FIX(git_tree_entrycount(tree));
 }
 
+static int rugged__treecount_cb(const char *root, const git_tree_entry *entry, void *data)
+{
+	struct rugged_treecount_cb_payload *payload = data;
+
+	if (payload->limit >= 0 && payload->count >= payload->limit)
+		return -1;
+	else if(git_tree_entry_type(entry) == GIT_OBJ_TREE)
+		return 0;
+	else {
+		++(payload->count);
+		return 1;
+	}
+}
```

Comment:
Hmm, good point. Rather than just counting blob, I could count all tree entries (including subtrees). This means that providing a limit will give protection from at least one more DOS vector. Would that cover the case you're thinking of?